mempool: Add the `bypass_{csv,cltv}` option to `testmempoolaccept`

w0xlt opened the pull request:
This PR adds the the `bypass_csv` and `bypass_cltv` options to `testmempoolaccept` RPC, making it possible to bypass CSV / CLTV execution.

For more context, check https://github.com/bitcoin/bitcoin/pull/25434#pullrequestreview-1020896809.

Since this PR is built on top of https://github.com/bitcoin/bitcoin/pull/25577, https://github.com/bitcoin/bitcoin/pull/25434 and https://github.com/bitcoin/bitcoin/pull/25532, only the last 4 commits are new here.

w0xlt committed:
mempool: change `maxfeerate` to be a members of an `options` argument

`maxfeerate` becomes a member of an "options" object rather than
a positional argument.

The idea is that any new parameters in the future will also go into
options.

w0xlt committed:
mempool: make `test_accept` parameter explicit

bvbfan commented:
One small note: multi-line comments are evil. Consider a valid code
```
/*
func(/*param*/ p);
```
comment captures entire function.
Consider
```
struct ProcessTxCriteria {
    bool test_accept;
    std::optional<MempoolTestCriteria> test_criteria;
} defaultTxCriterias = {};

ProcessTransaction(/*tx=*/txr, const ProcessTxCriteria& criterias = defaultTxCriterias);
```

w0xlt committed:
validation: Add `MemPoolBypass` for bypassing mempool checks

w0xlt committed:
mempool: add `MemPoolBypass` parameter to `ProcessNewPackage`

w0xlt committed:
mempool: add `MemPoolBypass` parameter to `AcceptToMemoryPool`

w0xlt committed:
mempool: add `MemPoolBypass` parameter to `ProcessTransaction`

w0xlt committed:
mempool: add `MemPoolBypass` parameter to `ATMPArgs`

w0xlt committed:
validation: add `m_bypass_{relative,absolute}_timelock`

w0xlt committed:
validation: add option to bypass contextual timelock checks

This is for test_accepts only, and not allowed in an actual submission
to mempool - see assert statements.

Provide an option to bypass BIP68 nSequence and nLockTime checks. This
means clients can use testmempoolaccept to check whether L2 transaction
chains (which typically have timelock conditions) are valid without
submitting them. Note that BIP112 and BIP65 are still checked since they
are script (non-contextual) checks. This does not invalidate any
signature or script caching.

Co-authored-by: glozow <gloriajzhao@gmail.com>

w0xlt committed:
rpc: add optional bypass_timelocks for testmempoolaccept

Co-authored-by: glozow <gloriajzhao@gmail.com>

w0xlt committed:
test: check if `bypass_{absolute,relative}_timelock` works

Test the bypass_timelock options in testmempoolaccept. This lets us
bypass BIP68 relative locktime checks (in nSequence) and absolute
locktime checks (in nLocktime).

Co-authored-by: glozow <gloriajzhao@gmail.com>

w0xlt committed:
test: check if bypass_timelocks do not affect scripts

OP_CSV and OP_CLTV script checks are still done,
so setting bypass_timelocks=True doesn't mean that
bad scripts pass.

Co-authored-by: glozow <gloriajzhao@gmail.com>

w0xlt committed:
validation: Add `bypass_feerate_accuracy` to `MempoolTestCriteria`

w0xlt committed:
validation: Add `bypass_feerate_accuracy` validation logic

w0xlt committed:
rpc: add `bypass_feerate_accuracy` option to `testmempoolaccept` RPC

w0xlt committed:
test: add functional tests for `bypass_feerate_accuracy` option

w0xlt committed:
validation: Add `m_bypass_{csv,cltv}` to `MempoolTestCriteria`

w0xlt committed:
validation: add logic to bypass CSV/CLTV execution

w0xlt committed:
rpc: add `bypass_{csv,cltv}` params to `testmempoolaccept` RPC

w0xlt committed:
test: add tests for `bypass_cltv` and `bypass_csv`

w0xlt commented:
Rebased on top of https://github.com/bitcoin/bitcoin/pull/25532/commits/972674cf316dc36f5881579331e159010b007129.

@bvbfan Good suggestion. A similar struct was implemented in https://github.com/bitcoin/bitcoin/pull/25577/commits/9e8703ca5ff0ce7a45f160ca26ac3ba7f1f71116

fanquake commented:
> Since this PR is built on top of https://github.com/bitcoin/bitcoin/pull/25577, https://github.com/bitcoin/bitcoin/pull/25434 and https://github.com/bitcoin/bitcoin/pull/25532,

I'm going to close this for now. I don't think there is a need to have this PR open yet, when it's building on multiple other PRs, which themselves still need conceptual review. You can always link to branches with example future changes, from the base PRs.
